### PR TITLE
Encode custom width integers using the next power of two width

### PR DESCRIPTION
**Pull request by LucasSte**
The Anchor IDL does not support custom width integers. Instead of creating a user-defined type for them, we will encode them using the smaller integer width greater than the given one.

### CODE REVIEW

File: src/codegen/encoding/borsh_encoding.rs [hunk 1]
```diff
@@ -235,17 +235,40 @@ impl BorshEncoding {
                 Expression::NumberLiteral(Loc::Codegen, Type::Uint(32), BigInt::from(1u8))
             }
 
-            Type::Uint(length) | Type::Int(length) => {
+            Type::Uint(width) | Type::Int(width) => {
+                let encoding_size = width.next_power_of_two();
+                let expr = if encoding_size != *width {
+                    if matches!(expr_ty, Type::Uint(_)) {
```
Comment (seanyoung) [inline src/codegen/encoding/borsh_encoding.rs:241]: We have `expr_ty.is_signed_int()` for this

**Review by seanyoung — APPROVED** [review]
How is this tested?

File: tests/borsh_encoding.rs [hunk 2]
```diff
@@ -106,96 +106,83 @@ impl BorshToken {
...
-        }
-        n if n <= 256 => {
-            let mut val = value.to_signed_bytes_le();
-            let byte_width = (width / 8) as usize;
-            match val.len().cmp(&byte_width) {
-                Ordering::Greater => {
-                    while val.len() > byte_width {
-                        val.pop();
-                    }
+    if width <= 8 {
```
Comment (xermicus) [inline tests/borsh_encoding.rs:109]: Just out of curiosity, why did you change this to be a `else if` chain?

**Review by xermicus — APPROVED** [review]

**Comment by LucasSte** [comment]
> How is this tested?

In `primitives.rs` we have tests for all bit-widths that are passed as arguments to functions, so the encoding and decoding is implicitly tested there: 

https://github.com/hyperledger/solang/blob/cfbc7a9f6bbd13fce0b5bee62313177e4fe61037/tests/solana_tests/primitives.rs#L354

File: tests/borsh_encoding.rs (hunk 2 above)
Comment (LucasSte) [reply tests/borsh_encoding.rs:109]: In the `match` statement, I wanted to match the width to an exact number, and now I have to deal with a range. If you know how to do this using `match`, please, let me know and I will change my code.

Comment (xermicus) [reply tests/borsh_encoding.rs:109]: [You can try something like this](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=986405b4429057f9e2af1a275ae283cc)

Comment (LucasSte) [reply tests/borsh_encoding.rs:109]: That looks nice! I'm going to change my code to match your suggestion.
